src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Stream.cs

```diff
@@ -164,17 +177,35 @@ public void OnResponseHeader(ReadOnlySpan<byte> name, ReadOnlySpan<byte> value)
                 }
             }
 
+            public void OnResponseHeadersStart()
+            {
+                lock (SyncObject)
+                {
```

Comment:
What other states can this be in other than ExpectingData?  Can we assert that it's only in one of those states?